core/src/test/java/com/linecorp/armeria/client/circuitbreaker/KeyedRetryConfigMappingTest.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.circuitbreaker;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.function.BiFunction;
+
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.retry.RetryConfig;
+import com.linecorp.armeria.client.retry.RetryConfigMapping;
+import com.linecorp.armeria.client.retry.RetryRule;
+import com.linecorp.armeria.client.retry.RetryRuleWithContent;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RpcResponse;
+
+class KeyedRetryConfigMappingTest {
+
+    @Test
+    void mapsCorrectly() throws Exception {
+        final BiFunction<ClientRequestContext, Request, String> keyFactory =
+                (ctx, req) -> ctx.endpoint().host() + '#' + ctx.path();
+        final BiFunction<ClientRequestContext, Request, RetryConfig<RpcResponse>> configFactory =
+                (ctx, req) -> {
+            if (ctx.endpoint().host().equals("host1")) {
+                return RetryConfig.<RpcResponse>builder(RetryRule.onException())
+                                  .maxTotalAttempts(1).responseTimeoutMillisForEachAttempt(1000).build();
+            } else if (ctx.endpoint().host().equals("host2")) {
+                if (ctx.path().equals("/path2")) {
+                    return RetryConfig.<RpcResponse>builder(
+                            RetryRuleWithContent.onResponse((c, r) -> completedFuture(true)))
+                            .maxTotalAttempts(2).responseTimeoutMillisForEachAttempt(2000).build();
+                } else {
+                    return RetryConfig.<RpcResponse>builder(RetryRule.onException())
+                                      .maxTotalAttempts(3).responseTimeoutMillisForEachAttempt(3000).build();
+                }
+            } else {
+                return RetryConfig.<RpcResponse>builder(
+                        RetryRuleWithContent.onResponse((c, r) -> completedFuture(false)))
+                        .maxTotalAttempts(4).responseTimeoutMillisForEachAttempt(4000).build();
+            }
+        };
```

Comment:
This API is flexible but not user-friendly a bit. How about providing `RetryConfigMappingBuilder` for building `RetryConfigMapping` fluently. 
For example:
```java
RetryConfigMapping.builder()
                  .host("foo.com", retryConfigForFoo)
                  .path("/my/path1", retryConfigForPath1)
                  .path("/my/path2", retryConfigForPath2)
                  .mapping((ctx, req) =>{
                      if(req.headers().get("SOME_HEADER") = "SOME_VALUE") {
                          return someRetryConfigForFoo;
                      }
                      return null;
                   })
                   .build(fallbackRetryConfig);
```

I think we don't need to cache `RetryConfig` by mapping the key. Let users decide whether to reuse it or create new one if they want to dynamically change the retry configuration.

What do you think? @haithamgabr

Reply:
I completely understand your concern. However, I spend a lot of time thinking about it from a user perspective, and I really think the current API is the best middle ground between flexibility and user-friendliness. The problem with an API like the one you suggested is that it will become super-confusing when multiple elements apply, like which config will be picked if the uri is foo.com/my/path1? both retryConfigForFoo and retryConfigForPath1 are eligible now. I really think going into this level of detailing will make for a very confusing API. I expanded the javadocs on RetryConfigMapping a lot to make it more detailed and provided an example. Notice that the example in this test case looks complicated because it's deliberately complicated (for rigorous testing). For the majority of users it will be much simpler, probably based only on one ctx element, but still users with more complicated needs have a way of filling them.

Regarding RetryConfig caching, I think it's important to cache because RetryConfig is actually stateful. The mapping is used between retries to obtain the retry rule from the config (RetryingClient lines 287 & 341). The retry rule holds a Backoff object which is stateful; returning a new backoff each retry will break the expected behavior as the docs warn:
https://armeria.dev/docs/client-retry#retryrule

Reply:
I tend to agree with @haithamgabr. One nit from myself is we could introduce a method like `builderForRpc()` and get rid of the type parameter in the call site.

Reply:
I made the parameterized builder() methods package private (to be used in other parameterized classes like AbstractRetryingCLientBuilder) and added public builderForRpc() and builderForHttp()

Reply:
>  I spent a lot of time thinking about it from a user perspective, and I really think the current API is the best middle ground between flexibility and user-friendliness.

I agree your opion and thoughts. 👍 Let's keep the current API design. 

>  The retry rule holds a Backoff object which is stateful.

JFYI, Backoff can be stateful. However, the default `Backoff`s offered by Armeria are not stateful. 😉 RetryClient saves the `STATE` to the attribute of a `ClientRequestContext`.